### PR TITLE
Magento product sync

### PR DESCRIPTION
**Pull request by adudekyeti**
Magento product sync

### CODE REVIEW

File: app/Integrations/Magento/Config.php [hunk 1]
```diff
@@ -19,23 +20,141 @@
...
 	 * @var self
 	 */
 	private static $instance;
 
 	/**
-	 * Function to get object to read configuration
+	 * Function to get object to read configuration.
 	 *
-	 * @return void
+	 * @return config
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Config.php]: ?

File: app/Integrations/Magento/Config.php [hunk 2]
```diff
@@ -19,23 +20,141 @@
...
+	 * @throws \yii\db\Exception
+	 */
+	public static function setLastScanId(string $type, string $name, int $id): void
+	{
+		$dbCommand = \App\Db::getInstance()->createCommand();
+		$data = [
+			'name' => "{$type}_last_scan_{$name}",
+			'value' => $id
+		];
+		if (!(new Query())->from('i_#__magento_config')
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Config.php]: move table name to const

File: app/Integrations/Magento/Config.php [hunk 3]
```diff
@@ -19,23 +20,141 @@
...
+	}
+
+	/**
+	 * Set start scan.
+	 *
+	 * @param string $type
+	 *
+	 * @throws \yii\db\Exception
+	 */
+	public static function setStartScan(string $type): void
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Config.php]: try to merge functions (setStartScan, setLastScanId)

File: app/Integrations/Magento/Config.php [hunk 4]
```diff
@@ -19,23 +20,141 @@
...
+	 * @return array
+	 */
+	public static function getLastScan(string $type): array
+	{
+		return [
+			'id' => self::getInstance()->get($type . '_last_scan_id') ?? 0,
+			'idcrm' => self::getInstance()->get($type . '_last_scan_idcrm') ?? 0,
+			'idmap' => self::getInstance()->get($type . '_last_scan_idmap') ?? 0,
+			'start_date' => self::getInstance()->get($type . '_start_scan_date') ?? false,
+			'end_date' => self::getInstance()->get($type . '_end_scan_date') ?? false,
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Config.php]: why you create so many self::getInstance() ?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: app/Integrations/Magento/Synchronizator/Record.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	 * Method to get which record have to update.
+	 *
+	 * @param array $recordYF
+	 * @param array $recordMagento
+	 *
+	 * @throws \ReflectionException
+	 *
+	 * @return bool|string
+	 */
+	public function whichToUpdate(array $recordYF, array $recordMagento)
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Record.php]: $recordYF, $recordMagento?

File: app/Integrations/Magento/Synchronizator/Record.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	{
+		$toUpdate = false;
+		$modifiedTimeYFEnd = strtotime($recordYF['modifiedtime']) > strtotime($this->lastScan['end_date']);
+		$modifiedTimeYFStart = strtotime($recordYF['modifiedtime']) < strtotime($this->lastScan['start_date']);
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Record.php]: for optimization

File: app/Integrations/Magento/Synchronizator/Record.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+{
+	/**
+	 * Format records id to given source.
+	 *
+	 * @param $ids
+	 * @param string $formatTo
+	 *
+	 * @return array
+	 */
+	public function getFormatedRecordsIds($ids, $formatTo = 'magento'): array
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Record.php]: ?

File: app/Integrations/Magento/Synchronizator/Base.php [hunk 8]
```diff
@@ -41,4 +75,114 @@ public function setConnector($connector)
 	 * @return void
 	 */
 	abstract public function process();
+
+	/**
+	 * Get record mapping.
+	 *
+	 * @param $type
+	 * @param bool $fromId
+	 * @param bool $limit
```
Comment (rskrzypczak) [inline app/Integrations/Magento/Synchronizator/Base.php]: ?

File: app/Integrations/Magento/Config.php [hunk 9]
```diff
@@ -19,23 +20,126 @@
...
+			];
+		}
+		if (!(new Query())->from(self::TABLE_NAME)
+			->where(['name' => $data['name']])
+			->exists()) {
+			$dbCommand->insert(self::TABLE_NAME, $data)->execute();
+		} else {
+			$dbCommand->update(self::TABLE_NAME, $data, ['name' => $data['name']])->execute();
+		}
+		static::updateData();
```
Comment (TomaszQr) [inline app/Integrations/Magento/Config.php]: $this->set($data['name'], data['value']);

File: app/Integrations/Magento/Config.php [hunk 10]
```diff
@@ -19,23 +20,126 @@
...
+			$data = (new Query())->select(['name', 'value'])->from(self::TABLE_NAME)->createCommand()->queryAllByGroup();
 			static::$instance->setData($data);
 		}
+		return static::$instance;
+	}
+
+	/**
+	 * Method to update data after update in db.
+	 */
+	public static function updateData(): void
```
Comment (TomaszQr) [inline app/Integrations/Magento/Config.php]: to remove

File: app/Integrations/Magento/Synchronizator/Product.php [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+	 * @param array $ids
+	 *
+	 * @throws \ReflectionException
+	 *
+	 * @return array
+	 */
+	public function getProductsCrm(array $ids = []): array
+	{
+		$queryGenerator = new \App\QueryGenerator('Products');
+		$query = $queryGenerator->createQuery();
```
Comment (TomaszQr) [inline app/Integrations/Magento/Synchronizator/Product.php]: fields ?

**Review by TomaszQr — CHANGES_REQUESTED** [review]

**Review by TomaszQr — APPROVED** [review]

File: app/Fields/File.php [hunk 12]
```diff
@@ -1409,4 +1409,33 @@ public static function saveFromBase($raw, $moduleName)
...
+
+	/**
+	 * Save file from given url.
+	 *
+	 * @param $url
+	 * @param $moduleName
+	 *
+	 * @return array
+	 */
+	public static function saveImageFromUrl($url, $moduleName): array
```
Comment (mariuszkrzaczkowski) [inline app/Fields/File.php]: ?

File: app/Integrations/Magento/Synchronizator/Base.php [hunk 13]
```diff
@@ -41,4 +111,241 @@ public function setConnector($connector)
...
+	 *
+	 * @return array
+	 */
+	public function parseShippingData(array $shippingData): array
+	{
+		$data = current($shippingData);
+		return [
+			'discountmode' => 1,
+			'taxmode' => 1,
+			'currency' => \App\Config::component('Magento', 'currencyId'),
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Base.php]: ?

File: app/Integrations/Magento/Synchronizator/Base.php [hunk 14]
```diff
@@ -41,4 +111,241 @@ public function setConnector($connector)
...
+			'name' => \App\Config::component('Magento', 'shippingServiceId'),
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Base.php]: ?

File: app/Integrations/Magento/Synchronizator/Base.php [hunk 15]
```diff
@@ -41,4 +111,241 @@ public function setConnector($connector)
...
+		if (method_exists($map, 'parseAdditionalData')) {
+			$additionalData = $map->parseAdditionalData($data);
+		}
+		return $additionalData;
+	}
+
+	/**
+	 * Get product sku map from YetiForce.
+	 */
+	public function getProductSkuMapCrm(): void
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Base.php]: ?

File: app/Integrations/Magento/Synchronizator/Integrators/Customer.php [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+ */
+
+namespace App\Integrations\Magento\Synchronizator\Integrators;
+
+/**
+ * Magento Product map class.
+ */
+abstract class Customer extends \App\Integrations\Magento\Synchronizator\Record
+{
+	public static $groups = [];
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Integrators/Customer.php:20]: ?

File: app/Integrations/Magento/Synchronizator/Maps/Customer.php [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+		'firstname' => 'firstname',
+		'lastname' => 'lastname',
+		'birthday' => 'dob',
+		'email' => 'email'
+	];
+
+	public static $additionalFields = [
+		'website_id' => ''
+	];
+
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Maps/Customer.php:37]: ?

File: app/Integrations/Magento/Synchronizator/Maps/Customer.php [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+	public function getCrmBirthday()
+	{
+		if (!isset($this->data['dob']) || '0000-00-00' === $this->data['dob']) {
+			return null;
+		}
+		return $this->data['dob'];
+	}
+
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Maps/Customer.php:45]: ?

File: config/Components/ConfigTemplates.php [hunk 19]
```diff
@@ -239,14 +240,140 @@
...
 			'description' => 'Password to account in magento.',
+			'validation' => function () {
+				return true;
+			}
 		],
 		'masterSource' => [
 			'default' => 'magento',
-			'description' => 'Set master source: yetiforce or magento',
+			'description' => 'Set master source: yetiforce or magento.',
+			'validation' => function () {
```
Comment (mariuszkrzaczkowski) [inline config/Components/ConfigTemplates.php:343]: ?

File: modules/Settings/Magento/views/Index.php [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+		$viewer = $this->getViewer($request);
+		$viewer->assign('MODULE', $request->getModule());
+		$viewer->assign('RECORD_MODEL', Settings_Magento_Record_Model::getCleanInstance());
+		$viewer->view('Index.tpl', $request->getModule(false));
+	}
+
+	public function getFooterScripts(App\Request $request)
+	{
+		return array_merge(parent::getFooterScripts($request), $this->checkAndConvertJsScripts([
+			"modules.Settings.{$request->getModule()}.resources.Edit",
```
Comment (mariuszkrzaczkowski) [inline modules/Settings/Magento/views/Index.php:28]: ?

File: app/Integrations/Magento/Synchronizator/Maps/Base.php [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+	}
+
+	/**
+	 * Return parsed data in YetiForce format.
+	 *
+	 * @param bool $onEdit
+	 *
+	 * @return array
+	 */
+	public function getDataCrm(bool $onEdit = false): array
```
Comment (mariuszkrzaczkowski) [inline app/Integrations/Magento/Synchronizator/Maps/Base.php:196]: ?

**Pull request merged**